README.md
# AI Text Transformer (Preview)

![Alt text](resx/icon.png) 

A Visual Studio Code extension that sends the selected text to an AI (e.g. ChatGPT) for transformation.

![Demo animation](docs/demonstration-1.gif)

:exclamation: Please bear in mind the following important notes

- **This extension is currently a preview app.**

- **The AI service may charge you for each transformation.**

- **All costs incurred by using this extension are the users own and not the responsibility of the developer.**

## Features

This very simple extension that passes any selected text in your document over to a chat AI along with instructions on how to improve it. You aren't limited in the ways you can improve the text, but some examples are making it:

- more casual
- more professional
- funnier

![Demo animation 2](docs/demonstration-2.gif)

The extension works by making a remote call to an AI service such as Open AI's ChatGPT.

Currently only one AI service is supported, OpenAI's ChatGPT. ChatGPT requires a paid account and an API key in order to make remote calls.

> Important: Open AI charges for each call to it's API. You, the user, are responsible for all costs and charges.

The user must supply their own API key. The cost of each call will depend on the number of tokens used in the request. This is determined by the following factors:

- the length of the selected text,
- the length of the full prompt,
- and the length of the chat response.

## Installation

- You must have an Open AI account in order to use this extension. If you don't have an account head to [platform.openai.com/signup](platform.openai.com/signup) and create a free account. If you already have an OpenAI account, simply log in.
- Next, click on your profile in the top-right corner and select "View API keys" from the drop-down menu.
- Here, click on "Create new secret key" and copy the API key. You can’t copy or view the entire API key later on, so please store the API key safely and securely. Do not share or display the API key in public. It’s a private key meant only for access to your account.

![Api keys](docs/open-ai-api-keys.webp)

- For extra security, we recommend setting spending limits on your OpenAI account in the Billing section [](https://platform.openai.com/account/billing/limits) of the OpenAI website.

![Usage limits](docs/usage-limits.png)

- Open up Visual Studio Code and head to "File | Preferences | Settings"
- In the `Search settings` box, type `nurdledurdle`.
- Now paste the API Key in the `nurdlydurdle: Api Key` setting.
- Close the Visual Studio Code settings.

## Usage

- In Visual Studio Code, select some text you want to change.
- Press `Ctrl+Shift+P` to open the quick command palette.
- Start typing `Transform selected text (AI)` and press Enter.

![Alt text](docs/command-palette.png)

- In the dialog box that appears, type the kind of change you'd like to make to the selected text.

![Alt text](docs/text-transform.png)

- Press enter, and follow further onscreen instructions.

## Extension Settings

This extension contributes the following settings:

- `aitexttransformer.apiKey`: The API Key used to call the AI Service
- `aitexttransformer.defaultFullPrompt`: The full prompt that is filled in and sent to the AI service.
- `aitexttransformer.defaultTransform`: The default text that is displayed to the user when suggesting new text improvements.
- `aitexttransformer.aiService`: The AI service to use. The only option currently available is `openAI`.

## Further information

Behind the scenes, the extension is simply sending a prompt to an AI Chat service and pasting in the response.

```plaintext
Make the text following the --- more casual. 

Return only the revised text without the --- in your response and nothing else.

---

SELECTED TEXT
```

The full prompt is customizable using the `aitexttransformer.defaultFullPrompt` setting.

## Known Issues

Please report any issues to the Github page.

## Release Notes

Below are release notes

### 0.0.1

Initial beta release.

---

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "nurdlydurdle" extension will be documented in this file.

## 0.0.1

- Initial beta release


out/quickOpen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.quickOpen = void 0;
const path = require("path");
const cp = require("child_process");
const vscode_1 = require("vscode");
const vscode_2 = require("vscode");
/**
 * A file opener using window.createQuickPick().
 *
 * It shows how the list of items can be dynamically updated based on
 * the user's input in the filter field.
 */
async function quickOpen() {
    const uri = await pickFile();
    if (uri) {
        const document = await vscode_2.workspace.openTextDocument(uri);
        await vscode_1.window.showTextDocument(document);
    }
}
exports.quickOpen = quickOpen;
class FileItem {
    constructor(base, uri) {
        this.base = base;
        this.uri = uri;
        this.label = path.basename(uri.fsPath);
        this.description = path.dirname(path.relative(base.fsPath, uri.fsPath));
    }
}
class MessageItem {
    constructor(base, message) {
        this.base = base;
        this.message = message;
        this.description = '';
        this.label = message.replace(/\r?\n/g, ' ');
        this.detail = base.fsPath;
    }
}
async function pickFile() {
    const disposables = [];
    try {
        return await new Promise((resolve, reject) => {
            const input = vscode_1.window.createQuickPick();
            input.placeholder = 'Type to search for files';
            let rgs = [];
            disposables.push(input.onDidChangeValue(value => {
                rgs.forEach(rg => rg.kill());
                if (!value) {
                    input.items = [];
                    return;
                }
                input.busy = true;
                const cwds = vscode_2.workspace.workspaceFolders ? vscode_2.workspace.workspaceFolders.map(f => f.uri.fsPath) : [process.cwd()];
                const q = process.platform === 'win32' ? '"' : '\'';
                rgs = cwds.map(cwd => {
                    const rg = cp.exec(`rg --files -g ${q}*${value}*${q}`, { cwd }, (err, stdout) => {
                        const i = rgs.indexOf(rg);
                        if (i !== -1) {
                            if (rgs.length === cwds.length) {
                                input.items = [];
                            }
                            if (!err) {
                                input.items = input.items.concat(stdout
                                    .split('\n').slice(0, 50)
                                    .map(relative => new FileItem(vscode_1.Uri.file(cwd), vscode_1.Uri.file(path.join(cwd, relative)))));
                            }
                            if (err && !err.killed && err.code !== 1 && err.message) {
                                input.items = input.items.concat([
                                    new MessageItem(vscode_1.Uri.file(cwd), err.message)
                                ]);
                            }
                            rgs.splice(i, 1);
                            if (!rgs.length) {
                                input.busy = false;
                            }
                        }
                    });
                    return rg;
                });
            }), input.onDidChangeSelection(items => {
                const item = items[0];
                if (item instanceof FileItem) {
                    resolve(item.uri);
                    input.hide();
                }
            }), input.onDidHide(() => {
                rgs.forEach(rg => rg.kill());
                resolve(undefined);
                input.dispose();
            }));
            input.show();
        });
    }
    finally {
        disposables.forEach(d => d.dispose());
    }
}
//# sourceMappingURL=quickOpen.js.map

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';

import { window, commands, ExtensionContext } from 'vscode';
import fetch from 'node-fetch';
//npm install node-fetch --sae-dev
//mpm install @types/node-fetch 
//npm uninstall node-fetch
//npm install -s node-fetch@2

import { StatusBarAlignment, StatusBarItem, Selection, workspace, TextEditor, ProgressLocation } from 'vscode';
import { Memento } from "vscode";

const MAX_TEXT_LENGTH = 1500;
const projectNameFriendly = "AI Text Transformer";
const projectName = "aitexttransformer";

function timeout(ms: number) {
	return new Promise(resolve => setTimeout(resolve, ms));
}

// /* See [here](https://www.chrishasz.com/blog/2020/07/28/vscode-how-to-use-local-storage-api/) to see how to store information across instances. */
// export class LocalStorageService {
    
//     constructor(private storage: Memento) { }   
    
//     public getValue<T>(key : string) : T{
//         return this.storage.get<T>(key, null);
//     }

//     public setValue<T>(key : string, value : T){
//         this.storage.update(key, value );
//     }
// }

async function callOpenAIMethod(prompt: string, apiKey: string) {


	
	const url = "https://api.openai.com/v1/chat/completions";

	const myData = {
		model: "gpt-3.5-turbo",
		messages: [
			{
				role: "user",
				content: prompt
			}
		]
	};

	const response = await fetch(url, {
		method: 'POST',
		headers: {
			// eslint-disable-next-line @typescript-eslint/naming-convention
			'Accept': 'application/json',
			// eslint-disable-next-line @typescript-eslint/naming-convention
			'Content-Type': 'application/json',
			// eslint-disable-next-line @typescript-eslint/naming-convention
			'Authorization': `Bearer ${apiKey}`,
		},
		body: JSON.stringify(myData),
	})
		;
	if (response.status === 200) {
		var json = await response.json();
		return json;
	}

	throw new Error(`Failed to speak to Open AI service. Response code ${response.status}`);

}

export async function callApiMethod(prompt: string, apiKey: string) {

	let retval = await window.withProgress({
		location: ProgressLocation.Notification,
		title: `${projectNameFriendly}\r\nCalling remote AI to improve selected text.\r\nPlease wait.`,
		cancellable: true
	}, async (progress, token) => {
		token.onCancellationRequested(() => {
			console.log("User canceled the operation");
		});
		//progress.report({ increment: 0 }); //Don't do this if it is indeterminate.
		const p = new Promise<{ success: Boolean, result: string, tokenCount: number }>(resolve => {
			callOpenAIMethod(prompt, apiKey).catch((reason) => {
				resolve({ success: false, result: "", tokenCount: 0 });
			}).then((val: any) => {
				const trimmedContent = val.choices[0].message.content.trim();
				resolve({ success: true, result: trimmedContent, tokenCount: val.usage.total_tokens });
			});
		});

		return p;
	});
	return retval;

}

async function showDialog(){
	//This is the only way to do a dialog,.
	await vscode.window
  .showInformationMessage("Do you want to do this?", "Yes", "No")
  .then(answer => {
    if (answer === "Yes") {
      // Run function
    }
  });
}

export function activate(context: vscode.ExtensionContext) {


	

	let disposable = vscode.commands.registerCommand(`${projectName}.transformSelectedText`, async () => {

		

		//Check the AI Service is supported. Otherwise, exit.
		const aiService = vscode.workspace.getConfiguration().get<string>(`${projectName}.aiService`);
		if (aiService !== "openAI") {
			vscode.window.showErrorMessage(`${projectNameFriendly}: AI Service \`${aiService}\` not supported. Try setting the \'${projectName}.aiService\` setting in the settings.json file.`);
			return;
		}

		if (!vscode.window.activeTextEditor) {
			vscode.window.showErrorMessage(`${projectNameFriendly}: Cannot run. There is no active document.`);
			return; // no editor
		}
		let editor = vscode.window.activeTextEditor;
		if (editor.selections.length > 1) {
			vscode.window.showErrorMessage(`${projectNameFriendly}: Cannot run. There are multiple selections.`);
			return;
		}
		if (editor.selections.length === 0) {
			vscode.window.showErrorMessage(`${projectNameFriendly}: Cannot run. Nothing is selected.`);
			return;
		}
		let selection = editor.selection;

		let { document } = editor;
		let selectedText = document.getText(selection);
		if (selectedText.length === 0) {
			vscode.window.showErrorMessage(`${projectNameFriendly}: Cannot run. Nothing is selected.`);
			return;
		}

		if (selectedText.length > MAX_TEXT_LENGTH) {
			vscode.window.showErrorMessage(`${projectNameFriendly}: Cannot run. The selected text length exceeds the maximum text length of ${MAX_TEXT_LENGTH} characters.`);
			return;
		}

		

		const defaultTransform = vscode.workspace.getConfiguration().get<string>(`${projectName}.defaultTransform`);

		const transformCommand = await window.showInputBox({
			prompt: "Describe how the selected text should be improved?\ne.g. more casual\r\n",
			value: defaultTransform,
			valueSelection: [0, defaultTransform?.length ?? 0],
			placeHolder: 'For example: more sarcastic OR funnier',
			validateInput: text => {
				//Check if the input is value.
				return null;
				// window.showInformationMessage(`Validating: ${text}`);
				// return text === '123' ? 'Not 123!' : null;
			}
		});
		if (!transformCommand) { return; }

		//Allow the capacity to skip confirmation. Don't expose this to the configuration system so that the way of discovering it is to read the code, thereby understanding the consequences of making the API calls.
		const secretSkipConfirmation = vscode.workspace.getConfiguration().get<boolean>(`${projectName}.skipConfirmation`);
		if (!secretSkipConfirmation) {
			const confirmationResult = await window.showQuickPick(['Yes', 'No'], {
				placeHolder: 'Are you sure? The AI service may charge you for this operation.',
				onDidSelectItem: item => { }
			});

			if (!confirmationResult || confirmationResult !== "Yes") {
				vscode.window.showInformationMessage(`${projectNameFriendly}: Execution cancelled.`);
				return;
			}
		}

		const defaultFullPrompt = vscode.workspace.getConfiguration().get<string>(`${projectName}.defaultFullPrompt`);
		let fullPrompt = <string>(defaultFullPrompt === null ? "" : defaultFullPrompt);
		fullPrompt = fullPrompt.replace("{0}", transformCommand);
		fullPrompt = fullPrompt.replace("{1}", selectedText);
		const apiKey = vscode.workspace.getConfiguration().get<string>(`${projectName}.apiKey`);

		if (!apiKey){
			vscode.window.showErrorMessage(`${projectNameFriendly}: No API key found in setting \`${projectName}.apiKey.\`. `);
			return;
		}

		const apiResult = await callApiMethod(fullPrompt, apiKey ?? "");
		if (!apiResult.success) {
			vscode.window.showErrorMessage(`${projectName}: Call to AI service failed.`);
			return;
		}

		window.showInformationMessage(`${projectNameFriendly}: A total of ${apiResult.tokenCount} tokens were used for this request.`);

		editor.edit(builder => {
			builder.replace(selection, apiResult.result);
		});

		//See [here](https://platform.openai.com/account/usage) for usage.
		//See [here](https://openai.com/pricing#image-models) for pricing. 
		//Currently: 	gpt-3.5-turbo	$0.002 / 1K tokens

	});

	context.subscriptions.push(disposable);
}

// This method is called when your extension is deactivated
export function deactivate() { }


out/basicInput.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.showInputBox = exports.showQuickPick = void 0;
const vscode_1 = require("vscode");
/**
 * Shows a pick list using window.showQuickPick().
 */
async function showQuickPick() {
    let i = 0;
    const result = await vscode_1.window.showQuickPick(['eins', 'zwei', 'drei'], {
        placeHolder: 'eins, zwei or drei',
        onDidSelectItem: item => vscode_1.window.showInformationMessage(`Focus ${++i}: ${item}`)
    });
    vscode_1.window.showInformationMessage(`Got: ${result}`);
}
exports.showQuickPick = showQuickPick;
/**
 * Shows an input box using window.showInputBox().
 */
async function showInputBox() {
    const result = await vscode_1.window.showInputBox({
        value: 'abcdef',
        valueSelection: [2, 4],
        placeHolder: 'For example: fedcba. But not: 123',
        validateInput: text => {
            vscode_1.window.showInformationMessage(`Validating: ${text}`);
            return text === '123' ? 'Not 123!' : null;
        }
    });
    vscode_1.window.showInformationMessage(`Got: ${result}`);
}
exports.showInputBox = showInputBox;
//# sourceMappingURL=basicInput.js.map

out/multiStepInput.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.multiStepInput = void 0;
const vscode_1 = require("vscode");
/**
 * A multi-step input using window.createQuickPick() and window.createInputBox().
 *
 * This first part uses the helper class `MultiStepInput` that wraps the API for the multi-step case.
 */
async function multiStepInput(context) {
    class MyButton {
        constructor(iconPath, tooltip) {
            this.iconPath = iconPath;
            this.tooltip = tooltip;
        }
    }
    const createResourceGroupButton = new MyButton({
        dark: vscode_1.Uri.file(context.asAbsolutePath('resources/dark/add.svg')),
        light: vscode_1.Uri.file(context.asAbsolutePath('resources/light/add.svg')),
    }, 'Create Resource Group');
    const resourceGroups = ['vscode-data-function', 'vscode-appservice-microservices', 'vscode-appservice-monitor', 'vscode-appservice-preview', 'vscode-appservice-prod']
        .map(label => ({ label }));
    async function collectInputs() {
        const state = {};
        await MultiStepInput.run(input => pickResourceGroup(input, state));
        return state;
    }
    const title = 'Create Application Service';
    async function pickResourceGroup(input, state) {
        const pick = await input.showQuickPick({
            title,
            step: 1,
            totalSteps: 3,
            placeholder: 'Pick a resource group',
            items: resourceGroups,
            activeItem: typeof state.resourceGroup !== 'string' ? state.resourceGroup : undefined,
            buttons: [createResourceGroupButton],
            shouldResume: shouldResume
        });
        if (pick instanceof MyButton) {
            return (input) => inputResourceGroupName(input, state);
        }
        state.resourceGroup = pick;
        return (input) => inputName(input, state);
    }
    async function inputResourceGroupName(input, state) {
        state.resourceGroup = await input.showInputBox({
            title,
            step: 2,
            totalSteps: 4,
            value: typeof state.resourceGroup === 'string' ? state.resourceGroup : '',
            prompt: 'Choose a unique name for the resource group',
            validate: validateNameIsUnique,
            shouldResume: shouldResume
        });
        return (input) => inputName(input, state);
    }
    async function inputName(input, state) {
        const additionalSteps = typeof state.resourceGroup === 'string' ? 1 : 0;
        // TODO: Remember current value when navigating back.
        state.name = await input.showInputBox({
            title,
            step: 2 + additionalSteps,
            totalSteps: 3 + additionalSteps,
            value: state.name || '',
            prompt: 'Choose a unique name for the Application Service',
            validate: validateNameIsUnique,
            shouldResume: shouldResume
        });
        return (input) => pickRuntime(input, state);
    }
    async function pickRuntime(input, state) {
        const additionalSteps = typeof state.resourceGroup === 'string' ? 1 : 0;
        const runtimes = await getAvailableRuntimes(state.resourceGroup, undefined /* TODO: token */);
        // TODO: Remember currently active item when navigating back.
        state.runtime = await input.showQuickPick({
            title,
            step: 3 + additionalSteps,
            totalSteps: 3 + additionalSteps,
            placeholder: 'Pick a runtime',
            items: runtimes,
            activeItem: state.runtime,
            shouldResume: shouldResume
        });
    }
    function shouldResume() {
        // Could show a notification with the option to resume.
        return new Promise((resolve, reject) => {
            // noop
        });
    }
    async function validateNameIsUnique(name) {
        // ...validate...
        await new Promise(resolve => setTimeout(resolve, 1000));
        return name === 'vscode' ? 'Name not unique' : undefined;
    }
    async function getAvailableRuntimes(resourceGroup, token) {
        // ...retrieve...
        await new Promise(resolve => setTimeout(resolve, 1000));
        return ['Node 8.9', 'Node 6.11', 'Node 4.5']
            .map(label => ({ label }));
    }
    const state = await collectInputs();
    vscode_1.window.showInformationMessage(`Creating Application Service '${state.name}'`);
}
exports.multiStepInput = multiStepInput;
// -------------------------------------------------------
// Helper code that wraps the API for the multi-step case.
// -------------------------------------------------------
class InputFlowAction {
}
InputFlowAction.back = new InputFlowAction();
InputFlowAction.cancel = new InputFlowAction();
InputFlowAction.resume = new InputFlowAction();
class MultiStepInput {
    constructor() {
        this.steps = [];
    }
    static async run(start) {
        const input = new MultiStepInput();
        return input.stepThrough(start);
    }
    async stepThrough(start) {
        let step = start;
        while (step) {
            this.steps.push(step);
            if (this.current) {
                this.current.enabled = false;
                this.current.busy = true;
            }
            try {
                step = await step(this);
            }
            catch (err) {
                if (err === InputFlowAction.back) {
                    this.steps.pop();
                    step = this.steps.pop();
                }
                else if (err === InputFlowAction.resume) {
                    step = this.steps.pop();
                }
                else if (err === InputFlowAction.cancel) {
                    step = undefined;
                }
                else {
                    throw err;
                }
            }
        }
        if (this.current) {
            this.current.dispose();
        }
    }
    async showQuickPick({ title, step, totalSteps, items, activeItem, ignoreFocusOut, placeholder, buttons, shouldResume }) {
        const disposables = [];
        try {
            return await new Promise((resolve, reject) => {
                const input = vscode_1.window.createQuickPick();
                input.title = title;
                input.step = step;
                input.totalSteps = totalSteps;
                input.ignoreFocusOut = ignoreFocusOut ?? false;
                input.placeholder = placeholder;
                input.items = items;
                if (activeItem) {
                    input.activeItems = [activeItem];
                }
                input.buttons = [
                    ...(this.steps.length > 1 ? [vscode_1.QuickInputButtons.Back] : []),
                    ...(buttons || [])
                ];
                disposables.push(input.onDidTriggerButton(item => {
                    if (item === vscode_1.QuickInputButtons.Back) {
                        reject(InputFlowAction.back);
                    }
                    else {
                        resolve(item);
                    }
                }), input.onDidChangeSelection(items => resolve(items[0])), input.onDidHide(() => {
                    (async () => {
                        reject(shouldResume && await shouldResume() ? InputFlowAction.resume : InputFlowAction.cancel);
                    })()
                        .catch(reject);
                }));
                if (this.current) {
                    this.current.dispose();
                }
                this.current = input;
                this.current.show();
            });
        }
        finally {
            disposables.forEach(d => d.dispose());
        }
    }
    async showInputBox({ title, step, totalSteps, value, prompt, validate, buttons, ignoreFocusOut, placeholder, shouldResume }) {
        const disposables = [];
        try {
            return await new Promise((resolve, reject) => {
                const input = vscode_1.window.createInputBox();
                input.title = title;
                input.step = step;
                input.totalSteps = totalSteps;
                input.value = value || '';
                input.prompt = prompt;
                input.ignoreFocusOut = ignoreFocusOut ?? false;
                input.placeholder = placeholder;
                input.buttons = [
                    ...(this.steps.length > 1 ? [vscode_1.QuickInputButtons.Back] : []),
                    ...(buttons || [])
                ];
                let validating = validate('');
                disposables.push(input.onDidTriggerButton(item => {
                    if (item === vscode_1.QuickInputButtons.Back) {
                        reject(InputFlowAction.back);
                    }
                    else {
                        resolve(item);
                    }
                }), input.onDidAccept(async () => {
                    const value = input.value;
                    input.enabled = false;
                    input.busy = true;
                    if (!(await validate(value))) {
                        resolve(value);
                    }
                    input.enabled = true;
                    input.busy = false;
                }), input.onDidChangeValue(async (text) => {
                    const current = validate(text);
                    validating = current;
                    const validationMessage = await current;
                    if (current === validating) {
                        input.validationMessage = validationMessage;
                    }
                }), input.onDidHide(() => {
                    (async () => {
                        reject(shouldResume && await shouldResume() ? InputFlowAction.resume : InputFlowAction.cancel);
                    })()
                        .catch(reject);
                }));
                if (this.current) {
                    this.current.dispose();
                }
                this.current = input;
                this.current.show();
            });
        }
        finally {
            disposables.forEach(d => d.dispose());
        }
    }
}
//# sourceMappingURL=multiStepInput.js.map

out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = exports.callApiMethod = void 0;
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require("vscode");
const vscode_1 = require("vscode");
const node_fetch_1 = require("node-fetch");
//npm install node-fetch --sae-dev
//mpm install @types/node-fetch 
//npm uninstall node-fetch
//npm install -s node-fetch@2
const vscode_2 = require("vscode");
const MAX_TEXT_LENGTH = 1500;
const projectNameFriendly = "AI Text Transformer";
const projectName = "aitexttransformer";
function timeout(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
// /* See [here](https://www.chrishasz.com/blog/2020/07/28/vscode-how-to-use-local-storage-api/) to see how to store information across instances. */
// export class LocalStorageService {
//     constructor(private storage: Memento) { }   
//     public getValue<T>(key : string) : T{
//         return this.storage.get<T>(key, null);
//     }
//     public setValue<T>(key : string, value : T){
//         this.storage.update(key, value );
//     }
// }
async function callOpenAIMethod(prompt, apiKey) {
    const url = "https://api.openai.com/v1/chat/completions";
    const myData = {
        model: "gpt-3.5-turbo",
        messages: [
            {
                role: "user",
                content: prompt
            }
        ]
    };
    const response = await (0, node_fetch_1.default)(url, {
        method: 'POST',
        headers: {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            'Accept': 'application/json',
            // eslint-disable-next-line @typescript-eslint/naming-convention
            'Content-Type': 'application/json',
            // eslint-disable-next-line @typescript-eslint/naming-convention
            'Authorization': `Bearer ${apiKey}`,
        },
        body: JSON.stringify(myData),
    });
    if (response.status === 200) {
        var json = await response.json();
        return json;
    }
    throw new Error(`Failed to speak to Open AI service. Response code ${response.status}`);
}
async function callApiMethod(prompt, apiKey) {
    let retval = await vscode_1.window.withProgress({
        location: vscode_2.ProgressLocation.Notification,
        title: `${projectNameFriendly}\r\nCalling remote AI to improve selected text.\r\nPlease wait.`,
        cancellable: true
    }, async (progress, token) => {
        token.onCancellationRequested(() => {
            console.log("User canceled the operation");
        });
        //progress.report({ increment: 0 }); //Don't do this if it is indeterminate.
        const p = new Promise(resolve => {
            callOpenAIMethod(prompt, apiKey).catch((reason) => {
                resolve({ success: false, result: "", tokenCount: 0 });
            }).then((val) => {
                const trimmedContent = val.choices[0].message.content.trim();
                resolve({ success: true, result: trimmedContent, tokenCount: val.usage.total_tokens });
            });
        });
        return p;
    });
    return retval;
}
exports.callApiMethod = callApiMethod;
async function showDialog() {
    //This is the only way to do a dialog,.
    await vscode.window
        .showInformationMessage("Do you want to do this?", "Yes", "No")
        .then(answer => {
        if (answer === "Yes") {
            // Run function
        }
    });
}
function activate(context) {
    let disposable = vscode.commands.registerCommand(`${projectName}.transformSelectedText`, async () => {
        //Check the AI Service is supported. Otherwise, exit.
        const aiService = vscode.workspace.getConfiguration().get(`${projectName}.aiService`);
        if (aiService !== "openAI") {
            vscode.window.showErrorMessage(`${projectNameFriendly}: AI Service \`${aiService}\` not supported. Try setting the \'${projectName}.aiService\` setting in the settings.json file.`);
            return;
        }
        if (!vscode.window.activeTextEditor) {
            vscode.window.showErrorMessage(`${projectNameFriendly}: Cannot run. There is no active document.`);
            return; // no editor
        }
        let editor = vscode.window.activeTextEditor;
        if (editor.selections.length > 1) {
            vscode.window.showErrorMessage(`${projectNameFriendly}: Cannot run. There are multiple selections.`);
            return;
        }
        if (editor.selections.length === 0) {
            vscode.window.showErrorMessage(`${projectNameFriendly}: Cannot run. Nothing is selected.`);
            return;
        }
        let selection = editor.selection;
        let { document } = editor;
        let selectedText = document.getText(selection);
        if (selectedText.length === 0) {
            vscode.window.showErrorMessage(`${projectNameFriendly}: Cannot run. Nothing is selected.`);
            return;
        }
        if (selectedText.length > MAX_TEXT_LENGTH) {
            vscode.window.showErrorMessage(`${projectNameFriendly}: Cannot run. The selected text length exceeds the maximum text length of ${MAX_TEXT_LENGTH} characters.`);
            return;
        }
        const defaultTransform = vscode.workspace.getConfiguration().get(`${projectName}.defaultTransform`);
        const transformCommand = await vscode_1.window.showInputBox({
            prompt: "Describe how the selected text should be improved?\ne.g. more casual\r\n",
            value: defaultTransform,
            valueSelection: [0, defaultTransform?.length ?? 0],
            placeHolder: 'For example: more sarcastic OR funnier',
            validateInput: text => {
                //Check if the input is value.
                return null;
                // window.showInformationMessage(`Validating: ${text}`);
                // return text === '123' ? 'Not 123!' : null;
            }
        });
        if (!transformCommand) {
            return;
        }
        //Allow the capacity to skip confirmation. Don't expose this to the configuration system so that the way of discovering it is to read the code, thereby understanding the consequences of making the API calls.
        const secretSkipConfirmation = vscode.workspace.getConfiguration().get(`${projectName}.skipConfirmation`);
        if (!secretSkipConfirmation) {
            const confirmationResult = await vscode_1.window.showQuickPick(['Yes', 'No'], {
                placeHolder: 'Are you sure? The AI service may charge you for this operation.',
                onDidSelectItem: item => { }
            });
            if (!confirmationResult || confirmationResult !== "Yes") {
                vscode.window.showInformationMessage(`${projectNameFriendly}: Execution cancelled.`);
                return;
            }
        }
        const defaultFullPrompt = vscode.workspace.getConfiguration().get(`${projectName}.defaultFullPrompt`);
        let fullPrompt = (defaultFullPrompt === null ? "" : defaultFullPrompt);
        fullPrompt = fullPrompt.replace("{0}", transformCommand);
        fullPrompt = fullPrompt.replace("{1}", selectedText);
        const apiKey = vscode.workspace.getConfiguration().get(`${projectName}.apiKey`);
        if (!apiKey) {
            vscode.window.showErrorMessage(`${projectNameFriendly}: No API key found in setting \`${projectName}.apiKey.\`. `);
            return;
        }
        const apiResult = await callApiMethod(fullPrompt, apiKey ?? "");
        if (!apiResult.success) {
            vscode.window.showErrorMessage(`${projectName}: Call to AI service failed.`);
            return;
        }
        vscode_1.window.showInformationMessage(`${projectNameFriendly}: A total of ${apiResult.tokenCount} tokens were used for this request.`);
        editor.edit(builder => {
            builder.replace(selection, apiResult.result);
        });
        //See [here](https://platform.openai.com/account/usage) for usage.
        //See [here](https://openai.com/pricing#image-models) for pricing. 
        //Currently: 	gpt-3.5-turbo	$0.002 / 1K tokens
    });
    context.subscriptions.push(disposable);
}
exports.activate = activate;
// This method is called when your extension is deactivated
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map